src/tactic/sort.lean

```diff
@@ -0,0 +1,168 @@
+    `[{ simp only [add_comm, add_assoc, add_left_comm], done, }] <|>
+    -- `[{ abel, done, }] <|> -- this works too. it's more robust but also a bit slower
+      fail format!"Failed to prove:\n {e_eq_fmt}",
+  h ← get_local n,
+  match hyp with
+  | some loc := do
+    hp ← get_local loc,
+    rewrite_hyp h hp,
+    tactic.clear h
+  | none     := do
+    rewrite_target h,
+    tactic.clear h
+  end
+| [] := skip
+end
+
+-- inductive sort_side | lhs | rhs | both
+
+/-- If the target is an equality, `sort_summands` sorts the summands on either side of the equality.
+-/
+meta def sort_summands (hyp : option name) (cmp_fn : expr → expr → bool) : tactic unit :=
+do
+  t ← match hyp with
+  | some hyp := do
+    hyp ← get_local hyp,
+    infer_type hyp
+  | none     := target
+  end,
+  match t.is_eq with
+  | none          := fail "The goal is not an equality"
+  | some (el, er) := do
+    sorted_sum_with_cmp_fn hyp cmp_fn el,
+    sorted_sum_with_cmp_fn hyp cmp_fn er
+  end
+
+end with_cmp_fn
+
+/--  The order on `polynomial.monomial n r`, where monomials are compared by their "exponent" `n`.
+If the expression is not a monomial, then the weight is `⊥`. -/
+meta def monomial_weight : expr → option ℕ
+| a := match a.app_fn with
+  | `(coe_fn $ polynomial.monomial %%n) := n.to_nat
+  | _ := none
+  end
+
+/--  The function we use to compare two `expr`:
+* all non-monomials are compared alphabetically;
+* all non-monomials are smaller than all monomials;
+* bigger monomials have higher exponent.
+-/
+meta def compare_fn (eₗ eᵣ : expr) : bool :=
+match (monomial_weight eₗ, monomial_weight eᵣ) with
+| (none, none)     := eₗ.to_string ≤ eᵣ.to_string -- this solution forces an unique ordering
+| (_, none)        := false
+| (none, _)        := true
+| (some l, some r) := l ≤ r
+end
+
+-- /--  If we have an expression involving monomials, `sum_sorted_monomials` returns an ordered sum
+-- of its terms.  Every summands that is not a monomial appears first, after that, monomials are
+-- sorted by increasing size of exponent. -/
+-- meta def sum_sorted_monomials (e : expr) : tactic unit :=
+-- sorted_sum_with_cmp_fn compare_fn e
+
+-- /--  If the target is an equality involving monomials,
+-- then  `sort_monomials_lhs` sorts the summands on the lhs. -/
+-- meta def sort_monomials_lhs : tactic unit :=
+-- sort_summands sort_side.lhs compare_fn
+
+-- /-- If the target is an equality involving monomials,
+-- then  `sort_monomials_rhs` sorts the summands on the rhs. -/
+-- meta def sort_monomials_rhs : tactic unit :=
+-- sort_summands sort_side.rhs compare_fn
+
+private meta def sort_monomials_core (allow_failure : bool) (hyp : option name) : tactic unit :=
+if allow_failure then sort_summands hyp compare_fn <|> skip else sort_summands hyp compare_fn
+
+/-- If the target is an equality involving monomials,
+then  `sort_monomials` sorts the summands on either side of the equality. -/
+meta def sort_monomials (locat : parse location) : itactic :=
```

Comment:
```suggestion
meta def sort_monomials (locat : parse location) : tactic unit :=
```

Reply:
Mario, thank you for your suggestion!

Honestly, I am not understanding much of what is going on with the implementation of the tactic and I'm happy that you have looked at it!